File: src/_cpp/TowerDefenseBase/game/WaveGenerator.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "WaveGenerator.h"
#include "gameboard.h"
#include "GameGS.h"
#include "Log.h"
NS_CC_BEGIN;

/*
samples:
<waves defaultdelayonewave="5">
    <wave index="0">
      <creep name="tank" delay="1" healthrate="1" score="10"/>
      <creep name="car" delay="0.5" healthrate="1" score="10"/>
      <creep name="car" delay="0.5" healthrate="1" score="10"/>
      <creep name="car" delay="0.5" healthrate="1" score="10"/>
      <creep name="car" delay="0.5" healthrate="1" score="10"/>
    </wave> 
    <wave index="0" defaultdelay="2.9" defaulthealthrate="1" defaultscore="10" defaultname="rotorplane0" count="5" />
</waves>
	
	*/

WaveGenerator::WaveGenerator()
: m_waveIndex()
, m_wavesCount()
, m_waves()
, m_currentWave()
, m_delayUnits()
, m_delayWaves()
, m_delayWavesNow(true)
, m_isRunning(true)
{
}

void WaveGenerator::load(const pugi::xml_node & node)
{
	m_delayWaves.set( node.attribute("defaultdelayonewave").as_float() );

	int index(0);
	FOR_EACHXML_BYTAG(node, waveXml, "wave")
	{
		m_waves.push_back(WaveInfo());
		WaveInfo & wave = m_waves.back();
		wave.index = index++;

		wave.type = strToUnitLayer( waveXml.attribute("routetype").as_string() );
		std::string dname = waveXml.attribute("defaultname").as_string();
		float dhealthRate = waveXml.attribute("defaulthealthrate").as_float();
		float dscore = waveXml.attribute("defaultscore").as_float();
		float ddelay = waveXml.attribute("defaultdelay").as_float();
		RouteSubType droutest = strToRouteSubType( waveXml.attribute("defaultroutesubtype").as_string() );
		int drouteindex = waveXml.attribute("defaultrouteindex").as_int();
		int count = waveXml.attribute("count").as_int();

		if ( count != 0 )
		{
			assert(dname.empty() == false );
			assert(dhealthRate != 0 );
			assert(dscore != 0 );
			assert(ddelay != 0 );
			while( count-- )
			{
				wave.creeps.push_back(dname);
				wave.healthRate.push_back(dhealthRate);
				wave.scores.push_back(dscore);
				wave.delayOneUnit.push_back(ddelay);
				wave.routeIndex.push_back( drouteindex );
				wave.routeSubType.push_back( droutest );
			}
		}
		else
		{
			FOR_EACHXML(waveXml, creep)
			{
				std::string name = creep.attribute("name").as_string();
				float healthRate = creep.attribute("healthrate").as_float();
				float score = creep.attribute("score").as_int();
				float delay = creep.attribute("delay").as_float();
				RouteSubType routest = strToRouteSubType( creep.attribute("routesubtype").value() );
				int routeindex = creep.attribute("routeindex").as_int();
				if ( name.empty() ) name = dname;
				if ( healthRate == 0 ) healthRate = dhealthRate;
				if ( score == 0 ) score = dscore;
				if ( delay == 0 ) delay = ddelay;
				if ( routest == RouteSubType::defaultvalue ) routest = droutest;
				if ( routeindex == 0 ) routeindex = drouteindex;
				wave.creeps.push_back(name);
				wave.healthRate.push_back(healthRate);
				wave.scores.push_back(score);
				wave.delayOneUnit.push_back(delay);
				wave.routeIndex.push_back( routeindex );
				wave.routeSubType.push_back( routest );
			}
		}
	}
}

void WaveGenerator::clear()
{
	m_waveIndex = 0;
	m_wavesCount = 0;

	m_waves.clear();
	m_currentWave = m_waves.end();
	
	m_delayUnits.set(0);
	m_delayWaves.set(0);
	
	m_delayWavesNow = false;
}

void WaveGenerator::start()
{
	m_delayUnits.reset();
	m_delayWaves.reset();
	m_currentWave = m_waves.end();
	
	m_delayUnits.set( 0 );

	m_waveIndex = 0;
	m_wavesCount = m_waves.size();
	m_delayWavesNow = false;
	GameGS::getInstance()->updateWaveCounter();
	resume();
}

void WaveGenerator::pause()
{
	m_isRunning = false;
}
void WaveGenerator::resume()
{
	m_isRunning = true;
}

void WaveGenerator::update(float dt)
{
	if( m_isRunning == false )
		return;

	if ( m_currentWave != m_waves.end() && m_currentWave->valid() == false )
	{
		m_waves.pop_front();
		m_currentWave = m_waves.end();
		onFinishWave();
	}
	else
	if ( m_currentWave != m_waves.end() && m_currentWave->valid() )
	{
		m_delayUnits.tick( dt );
		if( m_delayUnits )
		{
			generateCreep();
			if( m_currentWave->valid() )
			{
				float delay = m_currentWave->delayOneUnit.front();
				m_delayUnits.set( delay );
			}
			else
			{
				m_delayUnits.reset();
			}
		}
	}
	else
	{
		if( m_delayWavesNow == false )
		{
			m_delayWavesNow = true;
			if ( m_waves.empty() == false )
			{
				onPredelayWave( m_waves.front( ) );
			}
		}

		else
		{
			m_delayWaves.reset();
			m_delayWavesNow = false;
			if ( m_waves.empty() == false )
			{
				m_currentWave = m_waves.begin();
				assert( m_currentWave->valid() );
				onStartWave( *m_currentWave );
				float delay = m_currentWave->delayOneUnit.front();
				m_delayUnits.set( delay );
			}
		}
	}
}

void WaveGenerator::onPredelayWave( const WaveInfo & wave )
{
	GameGS::getInstance()->updateWaveCounter();
	GameGS::getInstance( )->getGameBoard( ).onPredelayWave( wave, (m_waveIndex != 0? m_delayWaves.value( ) : 0) );
}

void WaveGenerator::onStartWave( const WaveInfo & wave )
{
	m_waveIndex = std::min<unsigned>( m_waveIndex + 1, m_wavesCount );
	GameGS::getInstance( )->getGameBoard( ).onStartWave( wave );
}

void WaveGenerator::onFinishWave()
{
	GameGS::getInstance()->getGameBoard().onFinishWave();
	if ( m_waves.empty() )
	{
		onFinishAllWaves();
	}
}

void WaveGenerator::onFinishAllWaves()
{
	GameGS::getInstance()->getGameBoard().onFinishWaves();
}

void WaveGenerator::generateCreep()
{
	assert( m_currentWave != m_waves.end() );
	assert(m_currentWave->creeps.empty() == false );
	std::string name = m_currentWave->creeps.front();

	auto rst = m_currentWave->routeSubType.front();
	auto ri = m_currentWave->routeIndex.front();
	auto creep = GameGS::getInstance()->getGameBoard().createCreep(name, rst, ri);
	if ( creep )
	{
		unsigned cost = m_currentWave->scores.front();
		creep->setCost( cost );
		creep->setRate( m_currentWave->healthRate.front() );
	}
	m_currentWave->pop();
}


NS_CC_END;

File: src/_eu/game/WaveGenerator.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.WaveGenerator = {

    /** For Test Instance of */
    __WaveGenerator : true,

    /**@type {Integer} */
    m_waveIndex: 0,

    /**
     * @return {number}
     */
    getWaveIndex: function() {return this.m_waveIndex;},

    /**@type {Integer} */
    m_wavesCount: 0,

    /**
     * @return {number}
     */
    getWavesCount: function() {return this.m_waveIndex;},

    /** @type {Array< EU.WaveInfo >} */
    m_waves: [],

    /** @type {EU.WaveInfo} */
    m_currentWave: null,
    /** @type {Integer} */
    m_currentWaveIndex: null,

    /** @type {EU.Common.TimeCounter} */ m_delayUnits : null,
    /** @type {EU.Common.TimeCounter} */ m_delayWaves : null,
    /** @type {Boolean} */ m_delayWavesNow : true,
    /** @type {Boolean} */ m_isRunning : true,


    init: function(){
        this.m_currentWave = new EU.WaveInfo();
        this.m_delayUnits = new EU.Common.TimeCounter();
        this.m_delayWaves = new EU.Common.TimeCounter();
        this.m_currentWaveIndex = -1;
    },
    load: function(node)
    {
        this.m_delayWaves.set( EU.asObject(node.getAttribute("defaultdelayonewave"), 0.0) );

        var index = 0;
        for(var i=0; i < node.children.length; i++){
            var waveXml = node.children[i];
            if (waveXml.tagName == "wave") {

                this.m_waves.push(new EU.WaveInfo());
                var wave = this.m_waves[this.m_waves.length - 1];
                wave.index = index++;

                wave.type = EU.strToUnitLayer( EU.asObject(waveXml.getAttribute("routetype"), null) );
                var dname = EU.asObject(waveXml.getAttribute("defaultname"), null);
                var dhealthRate = EU.asObject(waveXml.getAttribute("defaulthealthrate"), 0.0);
                var dscore = EU.asObject(waveXml.getAttribute("defaultscore"), 0.0);
                var ddelay = EU.asObject(waveXml.getAttribute("defaultdelay"), 0.0);
                var droutest = EU.strToRouteSubType( EU.asObject(waveXml.getAttribute("defaultroutesubtype"), null) );
                var drouteindex = EU.asObject(waveXml.getAttribute("defaultrouteindex"), 0);
                var count = EU.asObject(waveXml.getAttribute("count"), 0);

                if ( count != 0 )
                {
                    EU.assert(dname ? dname.length > 0 : false );
                    EU.assert(dhealthRate != 0 );
                    EU.assert(dscore != 0 );
                    EU.assert(ddelay != 0 );
                    while( count-- )
                    {
                        wave.creeps.push(dname);
                        wave.healthRate.push(dhealthRate);
                        wave.scores.push(dscore);
                        wave.delayOneUnit.push(ddelay);
                        wave.routeIndex.push( drouteindex );
                        wave.routeSubType.push( droutest );
                    }
                }
                else
                {
                    for(var j=0; j < waveXml.children.length; j++){
                        var creep = waveXml.children[j];

                        var name = EU.asObject(creep.getAttribute("name"), null);
                        var healthRate = EU.asObject(creep.getAttribute("healthrate"), 0.0);
                        var score = EU.asObject(creep.getAttribute("score"), 0);
                        var delay = EU.asObject(creep.getAttribute("delay"), 0.0);
                        /**RouteSubType */
                        var routest = EU.strToRouteSubType( EU.asObject(creep.getAttribute("routesubtype") ));
                        var routeindex = EU.asObject(creep.getAttribute("routeindex"), 0);
                        if ( EU.xmlLoader.stringIsEmpty(name) ) name = dname;
                        if ( healthRate == 0 ) healthRate = dhealthRate;
                        if ( score == 0 ) score = dscore;
                        if ( delay == 0 ) delay = ddelay;
                        //TODO: EU.RouteSubType
                        if ( routest == EU.RouteSubType.defaultvalue ) routest = droutest;
                        if ( routeindex == 0 ) routeindex = drouteindex;
                        wave.creeps.push(name);
                        wave.healthRate.push(healthRate);
                        wave.scores.push(score);
                        wave.delayOneUnit.push(delay);
                        wave.routeIndex.push( routeindex );
                        wave.routeSubType.push( routest );
                    }
                }
            }
        }
    },

    clear: function()
    {
        this.m_waveIndex = 0;
        this.m_wavesCount = 0;
        
        while (this.m_waves.length > 0) { this.m_waves.pop(); }
        this.m_currentWave = null;
        this.m_currentWaveIndex = -1;

        this.m_delayUnits.set( 0 );
        this.m_delayWaves.set( 0 );

        this.m_delayWavesNow = false;
    },
    start: function()
    {
        this.m_delayUnits.reset();
        this.m_delayWaves.reset();
        this.m_currentWave = null;
        this.m_currentWaveIndex = -1;

        this.m_delayUnits.set( 0 );

        this.m_waveIndex = 0;
        this.m_wavesCount = this.m_waves.length;
        this.m_delayWavesNow = false;
        EU.GameGSInstance.updateWaveCounter();
        this.resume();
    },
    pause: function()
    {
        this.m_isRunning = false;
    },
    resume: function()
    {
        this.m_isRunning = true;
    },
    update: function(dt)
    {
        if( this.m_isRunning == false )
            return;

        if ( this.m_currentWave != null && this.m_currentWave.valid() == false )
        {
            this.m_waves.shift();
            this.m_currentWave = null;
            this.m_currentWaveIndex = -1;
            this.onFinishWave();
        }
        else
        if ( this.m_currentWave != null && this.m_currentWave.valid() )
        {
            this.m_delayUnits.tick( dt );
            if( this.m_delayUnits.is() )
            {
                this.generateCreep();
                if( this.m_currentWave.valid() )
                {
                    var delay = this.m_currentWave.delayOneUnit[0];
                    this.m_delayUnits.set( delay );
                }
                else
                {
                    this.m_delayUnits.reset();
                }
            }
        }
        else
        {
            if( this.m_delayWavesNow == false )
            {
                this.m_delayWavesNow = true;
                if ( this.m_waves.length > 0 )
                {
                    this.onPredelayWave( this.m_waves[0] );
                }
            }

            else
            {
                this.m_delayWaves.reset();
                this.m_delayWavesNow = false;
                if ( this.m_waves.length > 0 )
                {
                    this.m_currentWave = this.m_waves[0];
                    EU.assert( this.m_currentWave.valid() );
                    this.onStartWave(this.m_currentWave );
                    var delay = this.m_currentWave.delayOneUnit[0];
                    this.m_delayUnits.set( delay );
                }
            }
        }
    },

    onPredelayWave: function(wave )
    {
        EU.GameGSInstance.updateWaveCounter();
        EU.GameGSInstance.getGameBoard( ).onPredelayWave( wave, (this.m_waveIndex != 0? this.m_delayWaves.value() : 0) );
    },

    onStartWave: function(wave )
    {
        this.m_waveIndex = Math.min( this.m_waveIndex + 1, this.m_wavesCount );
        EU.GameGSInstance.getGameBoard( ).onStartWave( wave );
    },

    onFinishWave: function()
    {
        EU.GameGSInstance.getGameBoard().onFinishWave();
        if ( this.m_waves.length == 0 )
        {
            this.onFinishAllWaves();
        }
    },

    onFinishAllWaves: function()
    {
        EU.GameGSInstance.getGameBoard().onFinishWaves();
    },

    generateCreep: function()
    {
        EU.assert( this.m_currentWave != null );
        EU.assert(this.m_currentWave.creeps.length > 0);
        var name = this.m_currentWave.creeps[0];

        var rst = this.m_currentWave.routeSubType[0];
        var ri = this.m_currentWave.routeIndex[0];
        var creep = EU.GameGSInstance.getGameBoard().createCreep(name, rst, ri);
        if ( creep )
        {
            var cost = this.m_currentWave.scores[0];
            creep._cost = cost;
            creep.setRate( this.m_currentWave.healthRate[0] );
        }
        this.m_currentWave.pop();
    }
};


EU.WaveInfo = cc.Class.extend(
{
    /** @type {Array< string >} */creeps : null,
    /** @type {Array< var >} */delayOneUnit : null,
    /** @type {Array< Integer >} */scores : null,
    /** @type {Array< EU.RouteSubType >} */routeSubType : null,
    /** @type {Array< Integer >} */routeIndex : null,
    /** @type {Array< var >} */healthRate : null,
    /** @type {Array< EU.UnitLayer >} */type : null,
    /** @type {Integer} */index : -1,
    ctor: function()
    {
        "use strict";
        this.creeps = [];
        this.delayOneUnit = [];
        this.scores = [];
        this.routeSubType = [];
        this.routeIndex = [];
        this.healthRate = [];
        this.type = EU.UnitLayer.any;
    },
    valid: function()
    {
        "use strict";
        return this.delayOneUnit.length > 0;
    },
    pop: function()
    {
        "use strict";
        this.creeps.shift();
        this.delayOneUnit.shift();
        this.scores.shift();
        this.healthRate.shift();
        this.routeSubType.shift();
        this.routeIndex.shift();
    }
});

(function(){
    EU.WaveGenerator.init();
})();
